In [1]:
#Preprocessing Pipeline
import pandas as pd
df=pd.read_csv("/content/train.csv")

In [2]:
df=df.drop(["PassengerId","Name","Ticket"],axis=1)
df

Survived  Pclass     Sex   Age  SibSp  Parch     Fare Cabin Embarked
0           0       3    male  22.0      1      0   7.2500   NaN        S
1           1       1  female  38.0      1      0  71.2833   C85        C
2           1       3  female  26.0      0      0   7.9250   NaN        S
3           1       1  female  35.0      1      0  53.1000  C123        S
4           0       3    male  35.0      0      0   8.0500   NaN        S
..        ...     ...     ...   ...    ...    ...      ...   ...      ...
886         0       2    male  27.0      0      0  13.0000   NaN        S
887         1       1  female  19.0      0      0  30.0000   B42        S
888         0       3  female   NaN      1      2  23.4500   NaN        S
889         1       1    male  26.0      0      0  30.0000  C148        C
890         0       3    male  32.0      0      0   7.7500   NaN        Q

[891 rows x 9 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Survived  891 non-null    int64  
 1   Pclass    891 non-null    int64  
 2   Sex       891 non-null    object 
 3   Age       714 non-null    float64
 4   SibSp     891 non-null    int64  
 5   Parch     891 non-null    int64  
 6   Fare      891 non-null    float64
 7   Cabin     204 non-null    object 
 8   Embarked  889 non-null    object 
dtypes: float64(2), int64(4), object(3)
memory usage: 62.8+ KB


In [6]:
df["Age"]=df["Age"].fillna(df["Age"].median())

In [7]:
df["Embarked"]=df["Embarked"].fillna(df["Embarked"].mode())

In [8]:
df["Sex"]=df["Sex"].map({"male":0,"female":1})

In [10]:
df["Sex"].value_counts(dropna=False)

Sex
0    577
1    314
Name: count, dtype: int64

In [11]:
df=pd.get_dummies(df,columns=["Embarked"],drop_first=True)

In [12]:
df.columns

Index(['Survived', 'Pclass', 'Sex', 'Age', 'SibSp', 'Parch', 'Fare', 'Cabin',
       'Embarked_Q', 'Embarked_S'],
      dtype='object')

In [17]:
df=df.drop(["Cabin"],axis=1)

In [16]:
df.isnull().sum()
df.dtypes

Survived        int64
Pclass          int64
Sex             int64
Age           float64
SibSp           int64
Parch           int64
Fare          float64
Embarked_Q       bool
Embarked_S       bool
dtype: object

In [19]:
#Train Logistic Regression
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
x=df.drop(["Survived"],axis=1)
y=df["Survived"]
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)
lr=LogisticRegression(max_iter=1000)
lr.fit(x_train,y_train)
train_acc_lr=accuracy_score(y_train,lr.predict(x_train))
test_acc_lr=accuracy_score(y_test,lr.predict(x_test))
train_acc_lr,test_acc_lr

(0.800561797752809, 0.8100558659217877)

In [24]:
from sklearn.tree import DecisionTreeClassifier
dt=DecisionTreeClassifier(random_state=42)
dt.fit(x_train,y_train)
train_acc_dt=accuracy_score(y_train,dt.predict(x_train))
test_acc_dt=accuracy_score(y_test,dt.predict(x_test))
train_acc_dt,test_acc_dt

(0.9789325842696629, 0.7877094972067039)

Why Decision Trees overfit more

Decision Trees split the data repeatedly to perfectly classify training samples.

With no depth restriction, they can memorize training data, including noise.

This leads to:

Very high training accuracy

Lower test accuracy → poor generalization

Overfitting = high train acc, low test acc

In [26]:
#Simple Control
dt_limited=DecisionTreeClassifier(max_depth=3,random_state=42)
dt_limited.fit(x_train,y_train)
train_acc_dt_1=accuracy_score(y_train,dt_limited.predict(x_train))
test_acc_dt_1=accuracy_score(y_test,dt_limited.predict(x_test))
train_acc_dt_1,test_acc_dt_1

(0.8342696629213483, 0.7988826815642458)

Logistic Regression

Is a linear model

Learns global patterns

Has lower variance

Cannot memorize individual samples easily

Generalizes better on unseen data

Decision Tree

Is a non-linear model

Learns local rules

Has high variance

Easily memorizes data

Needs constraints (like max_depth)

Logistic Regression
Train acc is ~0.80056
Test acc is ~0.81006
Decision Tree
Train acc is ~0.978933
Test acc is ~0.787709

Decision Tree (depth=3)
Train acc is ~0.834270
Test acc is ~0.798883

Day 11 Summary:
Learned the difference between training and testing accuracy.
Understood how overfitting appears as high train accuracy but low test accuracy.
Observed that Decision Trees overfit without depth control.
Learned that limiting model complexity improves generalization.
Compared Logistic Regression and Decision Tree models and selected the model with better generalization.